src/main/kotlin/com/openlattice/graph/partioning/RepartitioningJob.kt

```diff
@@ -0,0 +1,352 @@
+        } finally {
+            publishJobState()
+        }
+    }
+
+    override fun updateProgress() {
+        progress = ((100 * (state.repartitionCount + state.deleteCount)) / state.needsMigrationCount).toByte()
+    }
+
+    private fun repartition(repartitionSql: String): Long = hds.connection.use { connection ->
+        progress = ((100 * state.repartitionCount) / state.needsMigrationCount).toByte()
+        try {
+            connection.prepareStatement(repartitionSql).use { repartitionData ->
+                bind(repartitionData)
+                repartitionData.executeLargeUpdate()
+            }
+        } catch (ex: Exception) {
+            abort { "Job $id terminated: ${ex.message}" }
+            throw ex
+        } finally {
+            publishJobState()
+        }
+    }
+
+    private fun getCount(countSql: String, partition: Int): Long = hds.connection.use { connection ->
+        try {
+            connection.prepareStatement(countSql).use { countQuery ->
+                bind(countQuery, partition)
+                val rs = countQuery.executeQuery()
+                rs.next()
+                ResultSetAdapters.count(rs)
+            }
+        } catch (ex: Exception) {
+            abort { "Job $id terminated: ${ex.message}" }
+            throw ex
+        } finally {
+            publishJobState()
+        }
+    }
+
+    private fun getNeedsMigrationCount(): Long = 4 * state.oldPartitions.fold(0L) { count, partition ->
+        count + getCount(idsNeedingMigrationCountSql, partition) +
+                getCount(dataNeedingMigrationCountSql, partition) +
+                getCount(edgesNeedingMigrationCountSql, partition)
+    }
+
+
+    private fun bind(ps: PreparedStatement, partition: Int = currentlyMigratingPartition) {
+        ps.setObject(1, state.entitySetId)
+        ps.setArray(2, PostgresArrays.createIntArray(ps.connection, state.newPartitions))
+        ps.setInt(3, partition)
+    }
+
+    private fun setPartitions(entitySetId: UUID, partitions: Set<Int>) {
+        require(entitySets.containsKey(entitySetId)) {
+            "Entity set $entitySetId not found"
+        }
+        entitySets.executeOnKey(entitySetId) {
+            val v = it.value
+            if (v != null) {
+                v.setPartitions(partitions)
+                it.setValue(v)
+            }
+        }
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is RepartitioningJob) return false
+        if (!super.equals(other)) return false
+
+        if (phase != other.phase) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = super.hashCode()
+        result = 31 * result + phase
+        return result
+    }
+
+}
+
+
+private val REPARTITION_SELECTOR = "partitions[ 1 + ((array_length(partitions,1) + (('x'||right(${ID.name}::text,8))::bit(32)::int % array_length(partitions,1))) % array_length(partitions,1))]"
```

Comment:
Seems ideal to have one central generating function for this that takes the id column name and maybe partitions column name too -- I realize this should never change, but I feel like we've copy/pasted most of this a bunch of times and seems like these could easily get out of sync

Reply:
This one is definitely different from all the previous versions as it can never return negative. Happy to move it somewhere else.